core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
+
+  /**
+   * Stack of context objects which is used to record attach & detach history on a thread.
+   */
+  private static final ThreadLocal<ArrayDeque<Context>> contextStack =
```

Comment:
We'd definitely be interested in seeing standardization occur, our motivations are highly aligned. Would it be fair to say that the conversation is still in it's infancy in terms of that occurring? I'll add some comments to 

https://github.com/DistributedTracing/continuation-local-storage-jvm/pull/1

1. Agreed this is not strictly required but it might become common practice.

2. That's certainly one case but I think immutability runs a little deeper than that.  Looking at your impl the save and restore methods don't protect from mutation of the values so it's possible for expectations to be violated as 'context' is shared among threads.

3. Agreed on the goal, I'll provide feedback on the expression

4. Noted, as you can see from the discussion here cancellation can be a property stored in context as opposed to a feature of it. That does make some of the listener interactions interesting but it's certainly doable.

For others reading along the alternate implementation is here
https://github.com/DistributedTracing/continuation-local-storage-jvm/pull/1